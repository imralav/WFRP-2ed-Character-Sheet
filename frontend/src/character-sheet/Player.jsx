import React from 'react';
import {FormattedMessage} from 'react-intl';

export default function Player(props) {
    const handleChange = event => {
       const changeData = {name: event.target.name, value: event.target.value};
       props.onChange(changeData);
    };
    return (
        <table>
            <thead>
            <tr>
                <th colSpan="2" className="header-font-size">
                    <FormattedMessage id="player" defaultMessage="Player"/>
                </th>
            </tr>
            </thead>
            <tbody>
            <tr>
                <td>
                    <div>
                        <label className="label-font-size">
                            <FormattedMessage id="name" defaultMessage="Name"/>:
                        </label>
                        <input type="text" value={props.data.name} name="name" onChange={event => handleChange(event)} className="content-font-size"/>
                    </div>
                </td>
                <td className="w-40p">
                    <div>
                        <label className="label-font-size">
                            <FormattedMessage id="game-master" defaultMessage="Game Master"/>:
                        </label>
                        <input type="text" value={props.data.gameMaster} name="gameMaster" onChange={event => handleChange(event)} className="content-font-size"/>
                    </div>
                </td>
            </tr>
            <tr>
                <td>
                    <div>
                        <label className="label-font-size">
                            <FormattedMessage id="campaign" defaultMessage="Campaign"/>:
                        </label>
                        <input type="text" value={props.data.campaign} name="campaign" onChange={event => handleChange(event)} className="content-font-size"/>
                    </div>
                </td>
                <td>
                    <div>
                        <label className="label-font-size">
                            <FormattedMessage id="campaign-year" defaultMessage="Campaign Year"/>:
                        </label>
                        <input type="text" value={props.data.campaignYear} name="campaignYear" onChange={event => handleChange(event)} className="content-font-size"/>
                    </div>
                </td>
            </tr>
            </tbody>
        </table>
    );
}