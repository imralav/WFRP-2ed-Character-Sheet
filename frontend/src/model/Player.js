import simplePropertyReplacementReducer from "./SimplePropertyReplacementReducer";

const DEFAULT_PLAYER = {name: '',gameMaster: '',campaign: '',campaignYear: ''};
export const UPDATE_PLAYER_ACTION = 'UPDATE_PLAYER';
export default simplePropertyReplacementReducer(UPDATE_PLAYER_ACTION, DEFAULT_PLAYER);