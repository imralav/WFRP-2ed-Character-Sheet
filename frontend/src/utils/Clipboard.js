export default class Clipboard {
    static fallbackCopyTextToClipboard(text) {
        let textArea = document.createElement("textarea");
        textArea.value = text;
        document.body.insertBefore(textArea, document.body.firstChild);
        textArea.focus();
        textArea.select();

        try {
            let successful = document.execCommand('copy');
            let msg = successful ? 'successful' : 'unsuccessful';
            console.log('Fallback: Copying text command was ' + msg);
        } catch (err) {
            console.error('Fallback: Oops, unable to copy', err);
        }
        document.body.removeChild(textArea);
    }

    static copyTextToClipboard(text) {
        Clipboard.fallbackCopyTextToClipboard(text);
        // if (!navigator.clipboard) {
        //     Clipboard.fallbackCopyTextToClipboard(text);
        //     return;
        // }
        //TODO: find out why the code below does not work
        // navigator.clipboard.writeText(text).then(() => {
        //     console.log('Async: Copying to clipboard was successful!');
        // }).catch(err => {
        //     console.error('Async: Could not copy text: ', err, ' falling back');
        //     Clipboard.fallbackCopyTextToClipboard(text);
        // });
    }
}