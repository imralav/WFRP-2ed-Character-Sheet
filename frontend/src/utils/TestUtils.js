import React, {Fragment} from 'react';

import en from 'react-intl/locale-data/en';
import {addLocaleData, IntlProvider} from 'react-intl';
import renderer from 'react-test-renderer';

import {BrowserRouter as Router, Route} from 'react-router-dom';

addLocaleData(en);

class ComponentBuilder {
    constructor(component) {
        this.component = component;
    }

    withIntl() {
        this.component = (
            <IntlProvider locale="en" messages={{}}>
                <Fragment>
                {this.component}
                </Fragment>
            </IntlProvider>
        );
        return this;
    }

    withRouter() {
        this.component = (
            <Router>
                <Fragment>
                    {this.component}
                </Fragment>
            </Router>
        );
        return this;
    }

    build() {
        return renderer.create(this.component);
    }
}

export default class TestUtils {
    static buildComponent(component) {
        return new ComponentBuilder(component);
    }
}