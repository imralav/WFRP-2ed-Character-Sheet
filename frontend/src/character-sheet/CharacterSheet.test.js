import React from 'react';
import CharacterSheet from '../character-sheet';
import Player from './Player';
import Clipboard from '../utils/Clipboard';
import TestUtils from '../utils/TestUtils';

describe('CharacterSheet', () => {
    let component;
    let componentRoot;
    let characterData;
    let tree;
    let clipboard;
    describe('setup', () => {
        describe('with no character data and no clipboard', () => {
            beforeEach(() => {
                characterData = {
                    player: {
                        name: 'playerName',
                        gameMaster: 'someGameMaster',
                        campaign: 'someCampaign',
                        campaignYear: 'someCampaignYear'
                    },
                    character: {
                        name: 'characterName'
                    }
                };
                clipboard = {
                    copyTextToClipboard: jest.fn()
                };
                component = TestUtils.buildComponent(<CharacterSheet/>).withIntl().build();
                componentRoot = component.root;
                tree = component.toJSON();
            });
            it('should match previous snapshot', () => {
                expect(tree).toMatchSnapshot();
            });
            it('should have correct clipboard set', () => {
                let characterSheet = componentRoot.findByType(CharacterSheet);
                expect(characterSheet.children[0].instance.clipboard).toBe(Clipboard);
            });
        });
        describe('with character data and clipboard', () => {
            let alertMock;
            beforeEach(() => {
                alertMock = {
                    show: jest.fn()
                };
                characterData = {
                    player: {
                        name: 'someName',
                        gameMaster: 'someGameMaster',
                        campaign: 'someCampaign',
                        campaignYear: 'someCampaignYear'
                    },
                    character: {
                        name: 'characterName'
                    },
                    experiencePoints: {
                        current: 50,
                        total: 150
                    }
                };
                alertMock = {
                    show: jest.fn()
                };
                clipboard = {
                    copyTextToClipboard: jest.fn()
                };
                component = TestUtils.buildComponent(<CharacterSheet characterData={characterData} clipboard={clipboard} alert={alertMock}/>).withIntl().build();
                componentRoot = component.root;
                tree = component.toJSON();
            });
            it('should match previous snapshot', () => {
                expect(tree).toMatchSnapshot();
            });
            describe('Copy character to clipboard', () => {
                beforeEach(() => {
                    let containerDiv = componentRoot.findByProps({id: 'character-json'});
                    let button = containerDiv.findByType('button');
                    button.props.onClick();
                });
                it('should copy encoded character data to clipboard', () => {
                    expect(clipboard.copyTextToClipboard).toHaveBeenCalledWith(btoa(JSON.stringify(characterData)));
                });
            });
            describe('Player data change', () => {
                const changeData = {name: 'name', value: 'someNewName'};
                let player;
                beforeEach(() => {
                    player = componentRoot.findByType(Player);
                    player.props.onChange(changeData);
                });
                it('should change state', () => {
                    let characterSheet = componentRoot.findByType(CharacterSheet);
                    expect(characterSheet.children[0].instance.state.characterData.player.name).toEqual(changeData.value);
                });
            });
        });
    });
});
