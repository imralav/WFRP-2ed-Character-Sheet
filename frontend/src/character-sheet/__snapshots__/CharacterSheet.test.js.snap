// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterSheet setup with character data and clipboard should match previous snapshot 1`] = `
<div
  className="row character-sheet"
>
  <div
    className="col-12"
    id="character-json"
  >
    <button
      className="btn btn-dark btn-sm content-font-size"
      onClick={[Function]}
    >
      <span>
        Copy character data to clipboard for later use
      </span>
    </button>
  </div>
  <div
    className="col-12"
    id="wfrp-logo-container"
  >
    <img
      alt="WFRP Logo"
      src="wfrp-logo.png"
    />
  </div>
  <div
    className="col-12"
    id="player"
  >
    <table>
      <thead>
        <tr>
          <th
            colSpan="2"
          >
            <span>
              Player
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div>
              <label>
                <span>
                  Name
                </span>
                :
              </label>
              <input
                name="name"
                onChange={[Function]}
                type="text"
                value="someName"
              />
            </div>
          </td>
          <td
            className="w-40p"
          >
            <div>
              <label>
                <span>
                  Game Master
                </span>
                :
              </label>
              <input
                name="gameMaster"
                onChange={[Function]}
                type="text"
                value="someGameMaster"
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div>
              <label>
                <span>
                  Campaign
                </span>
                :
              </label>
              <input
                name="campaign"
                onChange={[Function]}
                type="text"
                value="someCampaign"
              />
            </div>
          </td>
          <td>
            <div>
              <label>
                <span>
                  Campaign Year
                </span>
                :
              </label>
              <input
                name="campaignYear"
                onChange={[Function]}
                type="text"
                value="someCampaignYear"
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="character"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Character
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <span>
              Name
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Race
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Current Career
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Previous Careers
            </span>
            :
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="experience-points"
  >
    <table>
      <thead>
        <tr>
          <th
            colSpan="2"
          >
            <span>
              Experience Points
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <span>
              Current
            </span>
            :
          </td>
          <td>
            <span>
              Total
            </span>
            :
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="personal-details"
  >
    <table>
      <thead>
        <tr>
          <th
            colSpan="2"
          >
            <span>
              Personal Details
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <span>
              Age
            </span>
            :
          </td>
          <td>
            <span>
              Gender
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Eye Color
            </span>
            :
          </td>
          <td>
            <span>
              Weight
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Hair Color
            </span>
            :
          </td>
          <td>
            <span>
              Height
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Star Sign
            </span>
            :
          </td>
          <td>
            <span>
              Number of Siblings
            </span>
            :
          </td>
        </tr>
        <tr>
          <td
            colSpan="2"
          >
            <span>
              Birthplace
            </span>
            :
          </td>
        </tr>
        <tr>
          <td
            colSpan="2"
          >
            <span>
              Distinguishing Marks
            </span>
            :
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="character-profile"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Character Profile
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="combat-movement"
  >
    <table>
      <thead>
        <tr>
          <th
            colSpan="3"
          >
            <span>
              Combat Movement
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <span>
              Move/Disengage
            </span>
            :
          </td>
          <td>
            <span>
              Charge Attack
            </span>
            :
          </td>
          <td>
            <span>
              Run
            </span>
            :
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="weapons"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Weapons
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="armour"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Armour
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="armour-points"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Armour Points
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="skills"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Skills
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="talents"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Talents
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="trappings"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Trappings
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="money"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Money
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <span>
              Gold Crowns (gc)
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Silver Schillings (s)
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Brass Pennies (p)
            </span>
            :
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`CharacterSheet setup with no character data and no clipboard should match previous snapshot 1`] = `
<div
  className="row character-sheet"
>
  <div
    className="col-12"
    id="character-json"
  >
    <button
      className="btn btn-dark btn-sm content-font-size"
      onClick={[Function]}
    >
      <span>
        Copy character data to clipboard for later use
      </span>
    </button>
  </div>
  <div
    className="col-12"
    id="wfrp-logo-container"
  >
    <img
      alt="WFRP Logo"
      src="wfrp-logo.png"
    />
  </div>
  <div
    className="col-12"
    id="player"
  >
    <table>
      <thead>
        <tr>
          <th
            colSpan="2"
          >
            <span>
              Player
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div>
              <label>
                <span>
                  Name
                </span>
                :
              </label>
              <input
                name="name"
                onChange={[Function]}
                type="text"
                value={undefined}
              />
            </div>
          </td>
          <td
            className="w-40p"
          >
            <div>
              <label>
                <span>
                  Game Master
                </span>
                :
              </label>
              <input
                name="gameMaster"
                onChange={[Function]}
                type="text"
                value={undefined}
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div>
              <label>
                <span>
                  Campaign
                </span>
                :
              </label>
              <input
                name="campaign"
                onChange={[Function]}
                type="text"
                value={undefined}
              />
            </div>
          </td>
          <td>
            <div>
              <label>
                <span>
                  Campaign Year
                </span>
                :
              </label>
              <input
                name="campaignYear"
                onChange={[Function]}
                type="text"
                value={undefined}
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="character"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Character
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <span>
              Name
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Race
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Current Career
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Previous Careers
            </span>
            :
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="experience-points"
  >
    <table>
      <thead>
        <tr>
          <th
            colSpan="2"
          >
            <span>
              Experience Points
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <span>
              Current
            </span>
            :
          </td>
          <td>
            <span>
              Total
            </span>
            :
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="personal-details"
  >
    <table>
      <thead>
        <tr>
          <th
            colSpan="2"
          >
            <span>
              Personal Details
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <span>
              Age
            </span>
            :
          </td>
          <td>
            <span>
              Gender
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Eye Color
            </span>
            :
          </td>
          <td>
            <span>
              Weight
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Hair Color
            </span>
            :
          </td>
          <td>
            <span>
              Height
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Star Sign
            </span>
            :
          </td>
          <td>
            <span>
              Number of Siblings
            </span>
            :
          </td>
        </tr>
        <tr>
          <td
            colSpan="2"
          >
            <span>
              Birthplace
            </span>
            :
          </td>
        </tr>
        <tr>
          <td
            colSpan="2"
          >
            <span>
              Distinguishing Marks
            </span>
            :
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="character-profile"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Character Profile
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="combat-movement"
  >
    <table>
      <thead>
        <tr>
          <th
            colSpan="3"
          >
            <span>
              Combat Movement
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <span>
              Move/Disengage
            </span>
            :
          </td>
          <td>
            <span>
              Charge Attack
            </span>
            :
          </td>
          <td>
            <span>
              Run
            </span>
            :
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="col-12"
    id="weapons"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Weapons
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="armour"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Armour
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="armour-points"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Armour Points
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="skills"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Skills
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="talents"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Talents
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="trappings"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Trappings
            </span>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="col-12"
    id="money"
  >
    <table>
      <thead>
        <tr>
          <th>
            <span>
              Money
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <span>
              Gold Crowns (gc)
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Silver Schillings (s)
            </span>
            :
          </td>
        </tr>
        <tr>
          <td>
            <span>
              Brass Pennies (p)
            </span>
            :
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
